namespace as structure member name throws C parser off

**Test file:**

```
#ifdef	__cplusplus
extern "C" {
#endif

typedef struct foo_s {
	char *foo;
	char *namespace;
} foo_t;

typedef struct bar_s {
	char *bar;
} bar_t;

#ifdef	__cplusplus
}
#endif
```

**ctags version**:

```
$ ./ctags --version
Universal Ctags 0.0.0(1d3dbc8), Copyright (C) 2015 Universal Ctags Team
Universal Ctags is derived from Exuberant Ctags.
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Feb  7 2017, 13:35:30
  URL: https://ctags.io/
  Optional compiled features: +wildcards, +regex, +option-directory, +coproc, +xpath, +json, +interactive
```

**actual result from universal ctags**

```
!_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED	1	/0=unsorted, 1=sorted, 2=foldcase/
!_TAG_OUTPUT_MODE	u-ctags	/u-ctags or e-ctags/
!_TAG_PROGRAM_AUTHOR	Universal Ctags Team	//
!_TAG_PROGRAM_NAME	Universal Ctags	/Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL	https://ctags.io/	/official site/
!_TAG_PROGRAM_VERSION	0.0.0	/1d3dbc8/
foo	foo.h	/^	char *foo;$/;"	m	struct:foo_s	typeref:typename:char *
foo_s	foo.h	/^typedef struct foo_s {$/;"	s
```

**exuberant ctags**

```
Exuberant Ctags 5.8, Copyright (C) 1996-2009 Darren Hiebert
  Compiled: Apr 11 2012, 01:48:44
  Addresses: <dhiebert@users.sourceforge.net>, http://ctags.sourceforge.net
  Optional compiled features: +wildcards, +regex
```

**actual result from exuberant ctags**

```
!_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED	1	/0=unsorted, 1=sorted, 2=foldcase/
!_TAG_PROGRAM_AUTHOR	Darren Hiebert	/dhiebert@users.sourceforge.net/
!_TAG_PROGRAM_NAME	Exuberant Ctags	//
!_TAG_PROGRAM_URL	http://ctags.sourceforge.net	/official site/
!_TAG_PROGRAM_VERSION	5.8	//
bar	foo.h	/^	char *bar;$/;"	m	struct:bar_s
bar_s	foo.h	/^typedef struct bar_s {$/;"	s
bar_t	foo.h	/^} bar_t;$/;"	t	typeref:struct:bar_s
foo	foo.h	/^	char *foo;$/;"	m	struct:foo_s
foo_s	foo.h	/^typedef struct foo_s {$/;"	s
foo_t	foo.h	/^} foo_t;$/;"	t	typeref:struct:foo_s
```

```
[jet@localhost ctags]$ cat /tmp/foo.cpp
extern "C" {

struct s {
	char *namespace;
};

}
[jet@localhost ctags]$ gcc -c /tmp/foo.cpp
/tmp/foo.cpp:4:8: error: expected unqualified-id before ‘namespace’
  char *namespace;
        ^~~~~~~~~
/tmp/foo.cpp:4:7: error: expected ‘;’ at end of member declaration
  char *namespace;
       ^
/tmp/foo.cpp:4:8: error: expected unqualified-id before ‘namespace’
  char *namespace;
```

It seems that c++ compiler rejects code using a keyword as an indentifier even if in extern "C" block.

--- a/parsers/cxx/cxx_keyword.c
+++ b/parsers/cxx/cxx_keyword.c
@@ -9,6 +9,7 @@
 
 #include "cxx_keyword.h"
 #include "cxx_parser_internal.h"
+#include "cxx_debug.h"
 
 #include "keyword.h"
 
@@ -21,7 +22,10 @@ enum CXXKeywordFlag
 	// virtual, inline, friend, static
 	CXXKeywordExcludeFromTypeNames = (1 << 2),
 	// true, false, nullptr
-	CXXKeywordIsConstant = (1 << 3)
+	CXXKeywordIsConstant = (1 << 3),
+	// certain keywords are disabled "on-the-fly" to better
+	// handle C / C++ guessing errors (public, protected, private, namespace etc..)
+	CXXKeywordIsDisabled = (1 << 4)
 };
 
 typedef struct _CXXKeywordDescriptor
@@ -33,7 +37,7 @@ typedef struct _CXXKeywordDescriptor
 
 
 // This array is indexed by the CXXKeywordType enum
-static const CXXKeywordDescriptor g_aCXXKeywordTable[] = {
+static CXXKeywordDescriptor g_aCXXKeywordTable[] = {
 	{
 		"__attribute__",
 		CXXLanguageC | CXXLanguageCPP | CXXLanguageCUDA,
@@ -548,11 +552,57 @@ bool cxxKeywordIsConstant(CXXKeyword eKeywordId)
 			CXXKeywordIsConstant;
 }
 
+bool cxxKeywordIsCPPSpecific(CXXKeyword eKeywordId)
+{
+	return g_aCXXKeywordTable[eKeywordId].uLanguages == CXXLanguageCPP;
+}
+
 bool cxxKeywordExcludeFromTypeNames(CXXKeyword eKeywordId)
 {
 	return g_aCXXKeywordTable[eKeywordId].uFlags &
 			CXXKeywordExcludeFromTypeNames;
 }
+
+bool cxxKeywordIsDisabled(CXXKeyword eKeywordId)
+{
+	return g_aCXXKeywordTable[eKeywordId].uFlags &
+			CXXKeywordIsDisabled;
+}
+
+bool cxxKeywordEnablePublicProtectedPrivate(bool bEnableIt)
+{
+	bool bEnabledNow =
+			!(g_aCXXKeywordTable[CXXKeywordPUBLIC].uFlags & CXXKeywordIsDisabled);
+
+	if(bEnabledNow == bEnableIt)
+		return bEnabledNow;
+	
+	if(bEnableIt)
+	{
+		CXX_DEBUG_PRINT("Enabling public/protected/private keywords");
+
+		g_aCXXKeywordTable[CXXKeywordPUBLIC].uFlags &= ~CXXKeywordIsDisabled;
+		g_aCXXKeywordTable[CXXKeywordPROTECTED].uFlags &= ~CXXKeywordIsDisabled;
+		g_aCXXKeywordTable[CXXKeywordPRIVATE].uFlags &= ~CXXKeywordIsDisabled;
+	} else {
+		CXX_DEBUG_PRINT("Disabling public/protected/private keywords");
+
+		g_aCXXKeywordTable[CXXKeywordPUBLIC].uFlags |= CXXKeywordIsDisabled;
+		g_aCXXKeywordTable[CXXKeywordPROTECTED].uFlags |= CXXKeywordIsDisabled;
+		g_aCXXKeywordTable[CXXKeywordPRIVATE].uFlags |= CXXKeywordIsDisabled;
+	}
+
+	return bEnabledNow;
+}
+
+void cxxKeywordEnableFinal(bool bEnableIt)
+{
+	if(bEnableIt)
+		g_aCXXKeywordTable[CXXKeywordFINAL].uFlags &= ~CXXKeywordIsDisabled;
+	else
+		g_aCXXKeywordTable[CXXKeywordFINAL].uFlags |= CXXKeywordIsDisabled;
+}
+
 
 void cxxBuildKeywordHash(const langType eLangType,unsigned int uLanguage)
 {

--- a/parsers/cxx/cxx_keyword.h
+++ b/parsers/cxx/cxx_keyword.h
@@ -129,12 +129,56 @@ bool cxxKeywordIsConstant(CXXKeyword eKeywordId);
 bool cxxKeywordMayBePartOfTypeName(CXXKeyword eKeywordId);
 bool cxxKeywordIsTypeRefMarker(CXXKeyword eKeywordId);
 bool cxxKeywordExcludeFromTypeNames(CXXKeyword eKeywordId);
+bool cxxKeywordIsCPPSpecific(CXXKeyword eKeywordId);
+
 
 const char * cxxKeywordName(CXXKeyword eKeywordId);
 
 // uLanguage is really CXXLanguage, but we keep it as unsigned int to avoid
 // problems with header inclusions. It works anyway.
 void cxxBuildKeywordHash(const langType eLangType,unsigned int uLanguage);
+
+// Keyword enabled/disabled state management.
+//
+// public, protected, private, class, namespace... keywords are C++ only.
+// However when parsing .h files we don't know if they belong to a C program or C++
+// one and thus for safety we parse them as C++. If our guess is wrong then the parser
+// may become confused and in some cases even bail out.
+//
+// For this reason we enable/disable the processing of certain keyword sets
+// in certain contexts.
+
+
+//
+// "public,protected,private" keywords
+//
+// In header files we disable processing of such keywords until we either figure
+// out that the file really contains C++ or we start parsing a struct/union.
+//
+// This flag is meaningful only when parsing a .h file as C++ since in C
+// public/protected/private are never keywords and we assume that .cpp files
+// have C++ content (so public/protected/private are always keywords).
+//
+// This function returns the previous state of the public/protected/private keywords
+// enabled flag so it can be easily restored.
+bool cxxKeywordEnablePublicProtectedPrivate(bool bEnableIt);
+
+//
+// "namespace"
+//
+// This is not valid inside a class/struct/union
+//
+bool cxxKeywordEnableNamespace(bool bEnableIt);
+
+//
+// "final" keyword
+//
+// This is actually special at C++ level: it's a keyword only within a specific part
+// of a class declaration. In other contexts it's not a keyword.
+void cxxKeywordEnableFinal(bool bEnableIt);
+
+// Is the specific keyword currently disabled?
+bool cxxKeywordIsDisabled(CXXKeyword eKeywordId);
 
 
 #endif //!ctags_cxx_keyword_h_

--- a/parsers/cxx/cxx_keyword.h
+++ b/parsers/cxx/cxx_keyword.h
@@ -164,13 +164,6 @@ void cxxBuildKeywordHash(const langType eLangType,unsigned int uLanguage);
 bool cxxKeywordEnablePublicProtectedPrivate(bool bEnableIt);
 
 //
-// "namespace"
-//
-// This is not valid inside a class/struct/union
-//
-bool cxxKeywordEnableNamespace(bool bEnableIt);
-
-//
 // "final" keyword
 //
 // This is actually special at C++ level: it's a keyword only within a specific part

--- a/parsers/cxx/cxx_parser.c
+++ b/parsers/cxx/cxx_parser.c
@@ -795,7 +795,7 @@ static bool cxxParserParseClassStructOrUnionInternal(
 	// Skip attr and class-head-name
 
 	// enable "final" keyword handling
-	g_cxx.bParsingClassStructOrUnionDeclaration = true;
+	cxxKeywordEnableFinal(true);
 
 	unsigned int uTerminatorTypes = CXXTokenTypeEOF | CXXTokenTypeSingleColon |
 			CXXTokenTypeSemicolon | CXXTokenTypeOpeningBracket |
@@ -812,7 +812,7 @@ static bool cxxParserParseClassStructOrUnionInternal(
 
 		if(!bRet)
 		{
-			g_cxx.bParsingClassStructOrUnionDeclaration = false;
+			cxxKeywordEnableFinal(false);
 			CXX_DEBUG_LEAVE_TEXT("Could not parse class/struct/union name");
 			return false;
 		}
@@ -838,13 +838,14 @@ static bool cxxParserParseClassStructOrUnionInternal(
 
 		if(!bRet)
 		{
-			g_cxx.bParsingClassStructOrUnionDeclaration = false;
+			cxxKeywordEnableFinal(false);
 			CXX_DEBUG_LEAVE_TEXT("Could not parse class/struct/union name");
 			return false;
 		}
 	}
 
-	g_cxx.bParsingClassStructOrUnionDeclaration = false;
+	// Once we reached the terminator, "final" is not a keyword anymore.
+	cxxKeywordEnableFinal(false);
 
 	if(cxxTokenTypeIs(g_cxx.pToken,CXXTokenTypeParenthesisChain))
 	{
@@ -907,6 +908,30 @@ static bool cxxParserParseClassStructOrUnionInternal(
 			g_cxx.pTokenChain,
 			CXXTokenTypeIdentifier
 		);
+
+	// If no identifier has been found we can try some fallbacks.
+	if(!pClassName)
+	{
+		// If we're in C++ mode, but C++ language hasn't been confirmed yet,
+		// and there is a C++ specific keyword just before the terminator we found
+		// then we'll try to use it as class/struct/union name.
+		if(
+			cxxParserCurrentLanguageIsCPP() &&
+			(!g_cxx.bConfirmedCPPLanguage) &&
+			(eKeyword != CXXKeywordCLASS) &&
+			(g_cxx.pTokenChain->iCount >= 3) &&
+			cxxTokenTypeIs(g_cxx.pToken->pPrev,CXXTokenTypeKeyword) &&
+			cxxKeywordIsCPPSpecific(g_cxx.pToken->pPrev->eKeyword)
+		)
+		{
+			pClassName = g_cxx.pToken->pPrev;
+			pClassName->eType = CXXTokenTypeIdentifier;
+			CXX_DEBUG_PRINT(
+					"Found no class/struct/union name identifier but there is '%s' which might look good",
+					vStringValue(pClassName->pszWord)
+				);
+		}
+	}
 
 	int iPushedScopes = 0;
 
@@ -977,6 +1002,17 @@ static bool cxxParserParseClassStructOrUnionInternal(
 		cxxTokenChainDestroyLast(g_cxx.pTokenChain); // remove the {
 	} else {
 		cxxTokenChainClear(g_cxx.pTokenChain);
+	}
+
+	// OK. This seems to be a valid class/struct/union declaration.
+
+	if(
+			(uTagKind == CXXTagCPPKindCLASS) &&
+			(!g_cxx.bConfirmedCPPLanguage)
+		)
+	{
+		CXX_DEBUG_PRINT("Succeeded in parsing a C++ class: this really seems to be C++");
+		g_cxx.bConfirmedCPPLanguage = true;
 	}
 
 	tagEntryInfo * tag = cxxTagBegin(uTagKind,pClassName);
@@ -1092,24 +1128,17 @@ bool cxxParserParseClassStructOrUnion(
 	// Trick for "smart" handling of public/protected/private keywords in .h files parsed as C++.
 	// See the declaration of bEnablePublicProtectedPrivateKeywords for more info.
 
-	// Save the state
-	bool bEnablePublicProtectedPrivateKeywords = g_cxx.bEnablePublicProtectedPrivateKeywords;
-
-	// If parsing of the keywords was disabled, we're in C++ mode and the keyword is "class" then
-	// we're fairly certain that the source code is *really* C++.
-	if(
-			(!bEnablePublicProtectedPrivateKeywords) &&
-			(eKeyword == CXXKeywordCLASS) &&
-			cxxParserCurrentLanguageIsCPP()
-		)
-		bEnablePublicProtectedPrivateKeywords = true; // leave it on for good.
-
-	// Enable public/protected/private keywords
-	g_cxx.bEnablePublicProtectedPrivateKeywords = true;
+	// Enable public/protected/private keywords and save the previous state
+	bool bEnablePublicProtectedPrivateKeywords = cxxKeywordEnablePublicProtectedPrivate(true);
 
 	bool bRet = cxxParserParseClassStructOrUnionInternal(eKeyword,uTagKind,uScopeType);
 
-	g_cxx.bEnablePublicProtectedPrivateKeywords = bEnablePublicProtectedPrivateKeywords;
+	// If parsing succeeded, we're in C++ mode and the keyword is "class" then
+	// we're fairly certain that the source code is *really* C++.
+	if(g_cxx.bConfirmedCPPLanguage)
+		bEnablePublicProtectedPrivateKeywords = true; // leave it on for good: we're (almost) sure it's C++
+
+	cxxKeywordEnablePublicProtectedPrivate(bEnablePublicProtectedPrivateKeywords);
 
 	return bRet;
 }
@@ -1284,12 +1313,18 @@ bool cxxParserParseAccessSpecifier(void)
 			(eScopeType != CXXScopeTypeStruct)
 		)
 	{
-		// this is a syntax error: we're in the wrong scope.
 		CXX_DEBUG_LEAVE_TEXT(
-				"Access specified in wrong context (%d): "
-					"bailing out to avoid reporting broken structure",
+				"Access specified in wrong context (%d)",
 				eScopeType
 			);
+
+		if(!g_cxx.bConfirmedCPPLanguage)
+		{
+			CXX_DEBUG_LEAVE_TEXT("C++ language is not confirmed: we try to be flexible and not bail out");
+			return true;
+		}
+		// this is a syntax error: we're in the wrong scope.
+		CXX_DEBUG_LEAVE_TEXT("C++ language is confirmed: bailing out to avoid reporting broken structure");
 		return false;
 	}
 
@@ -1444,6 +1479,10 @@ rescanReason cxxCParserMain(const unsigned int passCount)
 {
 	CXX_DEBUG_ENTER();
 	cxxTagInitForLanguage(g_cxx.eCLangType);
+
+	g_cxx.bConfirmedCPPLanguage = false;
+	cxxKeywordEnablePublicProtectedPrivate(false);
+
 	rescanReason r = cxxParserMain(passCount);
 	CXX_DEBUG_LEAVE();
 	return r;
@@ -1453,6 +1492,11 @@ rescanReason cxxCUDAParserMain(const unsigned int passCount)
 {
 	CXX_DEBUG_ENTER();
 	cxxTagInitForLanguage(g_cxx.eCUDALangType);
+
+	// CUDA is C.
+	g_cxx.bConfirmedCPPLanguage = false;
+	cxxKeywordEnablePublicProtectedPrivate(false);
+
 	rescanReason r = cxxParserMain(passCount);
 	CXX_DEBUG_LEAVE();
 	return r;
@@ -1466,9 +1510,8 @@ rescanReason cxxCppParserMain(const unsigned int passCount)
 	// In header files we disable processing of public/protected/private keywords
 	// until we either figure out that this is really C++ or we're start parsing
 	// a struct/union.
-	g_cxx.bEnablePublicProtectedPrivateKeywords = !isInputHeaderFile();
-
-	CXX_DEBUG_PRINT("Parsing of public/protected/private is %d",g_cxx.bEnablePublicProtectedPrivateKeywords);
+	g_cxx.bConfirmedCPPLanguage = !isInputHeaderFile();
+	cxxKeywordEnablePublicProtectedPrivate(g_cxx.bConfirmedCPPLanguage);
 
 	rescanReason r = cxxParserMain(passCount);
 	CXX_DEBUG_LEAVE();

--- a/parsers/cxx/cxx_parser_block.c
+++ b/parsers/cxx/cxx_parser_block.c
@@ -205,17 +205,48 @@ process_token:
 				{
 					case CXXKeywordNAMESPACE:
 					{
-						if(cxxScopeGetType() == CXXScopeTypeNamespace ||
-						   cxxScopeGetType() == CXXScopeTypeFunction)
+						enum CXXScopeType eScopeType = cxxScopeGetType();
+						
+						if(
+								(
+									// toplevel or nested within a namespace
+									(eScopeType == CXXScopeTypeNamespace) ||
+									// namespace X = Y inside a function
+									(eScopeType == CXXScopeTypeFunction)
+								) && (
+									// either certainly C++
+									g_cxx.bConfirmedCPPLanguage ||
+									// or a "sane" namespace syntax
+									(
+										!cxxTokenChainPreviousTokenOfType(
+												g_cxx.pToken,
+												CXXTokenTypeStar |
+												CXXTokenTypeAnd |
+												CXXTokenTypeKeyword
+											)
+									)
+								)
+							)
 						{
-							// namespaces can be nested only within themselves
 							if(!cxxParserParseNamespace())
 							{
 								CXX_DEBUG_LEAVE_TEXT("Failed to parse namespace");
 								return false;
 							}
 						} else {
-							// hm... syntax error?
+							// If we're pretty sure this is C++ then this is a syntax error.
+							// If we're not sure (namely when we're in a *.h file) then
+							// let's try to be flexible: treat the namespace keyword as an identifier.
+							if(!g_cxx.bConfirmedCPPLanguage)
+							{
+								CXX_DEBUG_LEAVE_TEXT(
+									"Found namespace in unexpected place, but we're not sure it's really C++ "
+									"so we'll treat it as an identifier instead"
+								);
+								g_cxx.pToken->eType = CXXTokenTypeIdentifier;
+								continue;
+							}
+
 							CXX_DEBUG_LEAVE_TEXT(
 								"Found namespace in a wrong place: we're probably out of sync"
 							);

--- a/parsers/cxx/cxx_parser_internal.h
+++ b/parsers/cxx/cxx_parser_internal.h
@@ -297,25 +297,10 @@ typedef struct _CXXParserState
 	// Please note that the keywords appearing inside a () subchain are NOT marked.
 	unsigned int uKeywordState;
 
-	// This is used to handle the special case of "final" which is a keyword
-	// in class/struct/union declarations but not anywhere else
-	bool bParsingClassStructOrUnionDeclaration;
-
-	// public, protected and private keywords are C++ only.
-	// However when parsing .h files we don't know if they belong to
-	// a C program or C++ one and thus for safety we parse them as C++.
-	// If our guess is wrong then we might discard certain variable
-	// declarations thinking that they're composed of keywords.
-	//
-	// For public, protected, and private keywords we can try the following trick.
-	//
-	// In header files we disable processing of such keywords until we either figure
-	// out that the file really contains C++ or we start parsing a struct/union.
-	//
-	// This flag is meaningful only when parsing a .h file as C++ since in C
-	// public/protected/private are never keywords and we assume that .cpp files
-	// have C++ content (so public/protected/private are always keywords).
-	bool bEnablePublicProtectedPrivateKeywords;
+	// This is set to true when we're parsing a *.cpp file (cpp extension!)
+	// or we're parsing a header but we have encountered valid C++ constructs that
+	// definitely confirm we're parsing C++.
+	bool bConfirmedCPPLanguage;
 
 } CXXParserState;
 

--- a/parsers/cxx/cxx_parser_namespace.c
+++ b/parsers/cxx/cxx_parser_namespace.c
@@ -284,6 +284,14 @@ bool cxxParserParseNamespace(void)
 			"Should have an opening bracket here!"
 		);
 
+	if(!g_cxx.bConfirmedCPPLanguage)
+	{
+		CXX_DEBUG_PRINT(
+				"Succeeded in parsing a C++ namespace: this really seems to be C++"
+			);
+		g_cxx.bConfirmedCPPLanguage = true;
+	}
+
 	// Here we certainly got an opening bracket: namespace block
 
 	if(!cxxParserParseBlock(true))

--- a/parsers/cxx/cxx_parser_tokenizer.c
+++ b/parsers/cxx/cxx_parser_tokenizer.c
@@ -1167,19 +1167,7 @@ bool cxxParserParseNextToken(void)
 		int iCXXKeyword = lookupKeyword(t->pszWord->buffer,g_cxx.eLangType);
 		if(iCXXKeyword >= 0)
 		{
-			if(
-					(
-						(iCXXKeyword == CXXKeywordFINAL) &&
-						(!g_cxx.bParsingClassStructOrUnionDeclaration)
-					) || (
-						(
-							(iCXXKeyword == CXXKeywordPUBLIC) ||
-							(iCXXKeyword == CXXKeywordPROTECTED) ||
-							(iCXXKeyword == CXXKeywordPRIVATE)
-						) &&
-						(!g_cxx.bEnablePublicProtectedPrivateKeywords)
-					)
-				)
+			if(cxxKeywordIsDisabled((CXXKeyword)iCXXKeyword))
 			{
 				t->eType = CXXTokenTypeIdentifier;
 			} else {

--- a/parsers/cxx/cxx_parser_typedef.c
+++ b/parsers/cxx/cxx_parser_typedef.c
@@ -287,6 +287,7 @@ skip_to_comma_or_end:
 			// then we might try to look for a C++ identifier here.
 			if(
 				cxxParserCurrentLanguageIsCPP() &&
+				(!g_cxx.bConfirmedCPPLanguage) &&
 				cxxTokenTypeIs(pLookupStart,CXXTokenTypeKeyword) &&
 				cxxKeywordIsCPPSpecific(pLookupStart->eKeyword)
 			)

--- a/parsers/cxx/cxx_parser_typedef.c
+++ b/parsers/cxx/cxx_parser_typedef.c
@@ -257,6 +257,7 @@ skip_to_comma_or_end:
 		// Now look for the identifier
 		//
 		CXXTokenChain * pTParentChain;
+		CXXToken * pLookupStart = t;
 
 		if(cxxTokenTypeIs(t,CXXTokenTypeIdentifier))
 		{
@@ -281,8 +282,25 @@ skip_to_comma_or_end:
 
 		if(!t)
 		{
-			CXX_DEBUG_LEAVE_TEXT("Didn't find an identifier: something nasty is going on");
-			return;
+			// Not found yet.
+			// If we're in C++ mode but we haven't confirmed that the language is really C++
+			// then we might try to look for a C++ identifier here.
+			if(
+				cxxParserCurrentLanguageIsCPP() &&
+				cxxTokenTypeIs(pLookupStart,CXXTokenTypeKeyword) &&
+				cxxKeywordIsCPPSpecific(pLookupStart->eKeyword)
+			)
+			{
+				// treat as identifier
+				pLookupStart->eType = CXXTokenTypeIdentifier;
+				t = pLookupStart;
+			}
+
+			if(!t)
+			{
+				CXX_DEBUG_LEAVE_TEXT("Didn't find an identifier: something nasty is going on");
+				return;
+			}
 		}
 
 		tagEntryInfo * tag = cxxTagBegin(CXXTagKindTYPEDEF,t);

--- a/parsers/cxx/cxx_parser_using.c
+++ b/parsers/cxx/cxx_parser_using.c
@@ -153,6 +153,14 @@ bool cxxParserParseUsingClause(void)
 		}
 	}
 
+	if(!g_cxx.bConfirmedCPPLanguage)
+	{
+		CXX_DEBUG_PRINT(
+				"Succeeded in parsing C++ using: this really seems to be C++"
+			);
+		g_cxx.bConfirmedCPPLanguage = true;
+	}
+
 	CXX_DEBUG_LEAVE();
 	return true;
 }

